// tailwindcss-classnames/lib/cli/lib/non-configurable/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nonConfigurableClassNames = void 0;
var tslib_1 = require("tslib");
var accessibility_1 = tslib_1.__importDefault(require("./accessibility"));
var backgrounds_1 = tslib_1.__importDefault(require("./backgrounds"));
var borders_1 = tslib_1.__importDefault(require("./borders"));
var effects_1 = tslib_1.__importDefault(require("./effects"));
var filters_1 = tslib_1.__importDefault(require("./filters"));
var flexbox_1 = tslib_1.__importDefault(require("./flexbox"));
var grid_1 = tslib_1.__importDefault(require("./grid"));
var interactivity_1 = tslib_1.__importDefault(require("./interactivity"));
var layout_1 = tslib_1.__importDefault(require("./layout"));
var sizing_1 = tslib_1.__importDefault(require("./sizing"));
var spacing_1 = tslib_1.__importDefault(require("./spacing"));
var svg_1 = tslib_1.__importDefault(require("./svg"));
var tables_1 = tslib_1.__importDefault(require("./tables"));
var transforms_1 = tslib_1.__importDefault(require("./transforms"));
var transitionsAndAnimations_1 = tslib_1.__importDefault(require("./transitionsAndAnimations"));
var typography_1 = tslib_1.__importDefault(require("./typography"));
// These utility classes are classes that their names does not change. Their names are not
// configured by the `tailwind.config.js` file. e.g. 'flex-start', 'object-cover' etc.
// the only configuration that is applicable to these classes are pseudo-class variants.
exports.nonConfigurableClassNames = {
    accessibility: accessibility_1.default,
    backgrounds: backgrounds_1.default,
    borders: borders_1.default,
    effects: effects_1.default,
    filters: filters_1.default,
    flexBox: flexbox_1.default,
    grid: grid_1.default,
    interactivity: interactivity_1.default,
    layout: layout_1.default,
    sizing: sizing_1.default,
    spacing: spacing_1.default,
    svg: svg_1.default,
    tables: tables_1.default,
    transforms: transforms_1.default,
    transitionsAndAnimations: transitionsAndAnimations_1.default,
    typography: typography_1.default,
};
