// tailwindcss-classnames/lib/cli/lib/non-configurable/layout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var boxSizing = ['box-border', 'box-content'];
var display = [
    'block',
    'inline-block',
    'inline',
    'flex',
    'inline-flex',
    'grid',
    'inline-grid',
    'table',
    'inline-table',
    'table-row',
    'table-cell',
    'contents',
    'list-item',
    'hidden',
];
var boxDecorationBreak = ['decoration-slice', 'decoration-clone'];
var container = ['container'];
var float = ['float-right', 'float-left', 'float-none'];
var clear = ['clear-left', 'clear-right', 'clear-both', 'clear-none'];
var isolation = ['isolate', 'isolation-auto'];
var objectFit = [
    'object-contain',
    'object-cover',
    'object-fill',
    'object-none',
    'object-scale-down',
];
var overflow = [
    'overflow-auto',
    'overflow-hidden',
    'overflow-visible',
    'overflow-scroll',
    'overflow-x-auto',
    'overflow-y-auto',
    'overflow-x-hidden',
    'overflow-y-hidden',
    'overflow-x-visible',
    'overflow-y-visible',
    'overflow-x-scroll',
    'overflow-y-scroll',
];
var overscrollBehavior = [
    'overscroll-auto',
    'overscroll-contain',
    'overscroll-none',
    'overscroll-y-auto',
    'overscroll-y-contain',
    'overscroll-y-none',
    'overscroll-x-auto',
    'overscroll-x-contain',
    'overscroll-x-none',
];
var position = ['static', 'fixed', 'absolute', 'relative', 'sticky'];
var visibility = ['visible', 'invisible'];
exports.default = {
    display: display,
    boxDecorationBreak: boxDecorationBreak,
    boxSizing: boxSizing,
    container: container,
    float: float,
    clear: clear,
    isolation: isolation,
    objectFit: objectFit,
    overflow: overflow,
    overscrollBehavior: overscrollBehavior,
    position: position,
    visibility: visibility,
};
