// tailwindcss-classnames/es/cli/lib/non-configurable/flexbox.js
var flexDirection = ['flex-row', 'flex-row-reverse', 'flex-col', 'flex-col-reverse'];
var flexWrap = ['flex-nowrap', 'flex-wrap', 'flex-wrap-reverse'];
var alignItems = ['items-stretch', 'items-start', 'items-center', 'items-end', 'items-baseline'];
var alignContent = [
    'content-start',
    'content-center',
    'content-end',
    'content-between',
    'content-around',
];
var alignSelf = ['self-auto', 'self-start', 'self-center', 'self-end', 'self-stretch', 'self-baseline'];
var placeContent = [
    'place-content-start',
    'place-content-center',
    'place-content-end',
    'place-content-between',
    'place-content-around',
    'place-content-evenly',
    'place-content-stretch',
];
var placeItems = [
    'place-items-auto',
    'place-items-start',
    'place-items-center',
    'place-items-end',
    'place-items-stretch',
];
var placeSelf = [
    'place-self-auto',
    'place-self-start',
    'place-self-center',
    'place-self-end',
    'place-self-stretch',
];
var justifyContent = [
    'justify-start',
    'justify-center',
    'justify-end',
    'justify-between',
    'justify-around',
    'justify-evenly',
];
var justifyItems = [
    'justify-items-auto',
    'justify-items-start',
    'justify-items-center',
    'justify-items-end',
    'justify-items-stretch',
];
var justifySelf = [
    'justify-self-auto',
    'justify-self-start',
    'justify-self-center',
    'justify-self-end',
    'justify-self-stretch',
];
var flex = ['flex-initial', 'flex-1', 'flex-auto', 'flex-none'];
export default {
    flexDirection: flexDirection,
    flexWrap: flexWrap,
    alignItems: alignItems,
    alignContent: alignContent,
    alignSelf: alignSelf,
    placeContent: placeContent,
    placeItems: placeItems,
    placeSelf: placeSelf,
    justifyContent: justifyContent,
    justifyItems: justifyItems,
    justifySelf: justifySelf,
    flex: flex,
};
