// tailwindcss-classnames/es/cli/lib/non-configurable/typography.js
var fontSmoothing = ['antialiased', 'subpixel-antialiased'];
var fontStyle = ['italic', 'non-italic'];
var fontVariantNumeric = [
    'normal-nums',
    'ordinal',
    'slashed-zero',
    'lining-nums',
    'oldstyle-nums',
    'proportional-nums',
    'tabular-nums',
    'diagonal-fractions',
    'stacked-fractions',
];
var listStylePosition = ['list-inside', 'list-outside'];
var textAlign = ['text-left', 'text-center', 'text-right', 'text-justify'];
var textDecoration = ['underline', 'line-through', 'no-underline'];
var textTransform = ['uppercase', 'lowercase', 'capitalize', 'normal-case'];
var textOverflow = ['truncate', 'overflow-ellipsis', 'overflow-clip'];
var verticalAlign = [
    'align-baseline',
    'align-top',
    'align-middle',
    'align-bottom',
    'align-text-top',
    'align-text-bottom',
];
var whitespace = [
    'whitespace-normal',
    'whitespace-nowrap',
    'whitespace-pre',
    'whitespace-pre-line',
    'whitespace-pre-wrap',
];
var wordBreak = ['break-normal', 'break-words', 'break-all'];
export default {
    fontSmoothing: fontSmoothing,
    fontStyle: fontStyle,
    fontVariantNumeric: fontVariantNumeric,
    listStylePosition: listStylePosition,
    textAlign: textAlign,
    textDecoration: textDecoration,
    textTransform: textTransform,
    textOverflow: textOverflow,
    verticalAlign: verticalAlign,
    whitespace: whitespace,
    wordBreak: wordBreak,
};
